pythonforandroid/graph.py

```diff
@@ -99,17 +129,125 @@ def find_order(graph):
                 bset.discard(result)
 
 
-def get_recipe_order_and_bootstrap(ctx, names, bs=None):
-    recipes_to_load = set(names)
+def obvious_conflict_checker(ctx, name_tuples, blacklist=[]):
+    """ This is a pre-flight check function that will completely ignore
+        recipe order or choosing an actual value in any of the multiple
+        choice tuples/dependencies, and just do a very basic obvious
+        conflict check.
+    """
+    deps_were_added_by = dict()
+    deps = set()
+
+    # Add dependencies for all recipes:
+    to_be_added = [(name_tuple, None) for name_tuple in name_tuples]
+    while len(to_be_added) > 0:
+        current_to_be_added = list(to_be_added)
+        to_be_added = []
+        for (added_tuple, adding_recipe) in current_to_be_added:
+            if len(added_tuple) > 1:
+                # No obvious commitment in what to add, ignore it.
+                # (Remember this function only catches obvious issues)
+                continue
+
+            name = added_tuple[0]
+            recipe_conflicts = []
+            recipe_dependencies = []
+            try:
+                # Get recipe to add and who's ultimately adding it:
+                recipe = Recipe.get_recipe(name, ctx)
+                recipe_conflicts = [c.lower() for c in recipe.conflicts]
+                recipe_dependencies = \
+                    get_dependency_tuple_list_for_recipe(
+                        recipe, blacklist=blacklist
+                    )
+            except ValueError:
+                pass
+            adder_recipe_1_name = adding_recipe or name
+
+            # See if this recipe itself conflicts with what was added before:
+            triggered_conflicts = []
+            for dep_tuple_list in deps:
+                if len(set(dep_tuple_list).intersection(
+                        set(recipe_conflicts))) == len(dep_tuple_list):
+                    triggered_conflicts.append(dep_tuple_list)
+
+            # See if what was added before conflicts with this recipe:
+            for dep_tuple_list in deps:
+                if len(dep_tuple_list) > 1:
+                    # Not an obvious commitment, so we won't check.
+                    # (remember this function only catches obvious issues)
+                    continue
+                try:
+                    conflicts = [c.lower() for c in
+                                 Recipe.get_recipe(dep_tuple_list[0], ctx).
+                                 conflicts]
+                    if name in conflicts:
+                        triggered_conflicts.append(dep_tuple_list)
+                except ValueError:
+                    pass
```

Comment:
No. Check beforehand instead of evaluating and waiting for exception.

Reply:
I actually like it. Isn't Python usually pushing for `EAFP` https://docs.python.org/3/glossary.html ?
But that's up to you guys :+1:

Reply:
It's in 3 loops and calls a loop. Any such code shape will slow everything down. Check for what you expect to be wrong and then move to the other iteration.

Reply:
this is really just looking so weird because I didn't assign `Recipe.get_recipe` outside of the loop which @KeyWeeUsr already reasonably suggested. With that changed this will be implicitly fixed :+1:

Reply:
changed it now :+1: